File: ocaml/list-ops/list_ops.ml
let rec fold ~(init: 'acc) ~(f: 'acc -> 'a -> 'acc) xs =
  match xs with
  | [] -> init
  | x :: xs -> fold ~init:(f init x) ~f xs

let reverse xs =
  fold ~init:[] ~f:(fun acc x -> x :: acc) xs

let length xs =
  fold ~init:0 ~f:(fun acc _ -> acc + 1) xs

let append xs ys =
  fold ~init:ys ~f:(fun acc x -> x :: acc) (reverse xs)

let concat xs =
  fold ~init:[] ~f:(fun acc x -> append x acc) (reverse xs)

let filter ~(f: 'a -> bool) xs =
  reverse (fold ~init:[] ~f:(fun acc x -> if f x then x :: acc else acc) xs)

let rec map ~(f: 'a -> 'b) xs =
  match xs with
  | [] -> []
  | x :: xs -> f x :: map ~f xs

File: ruby/list-ops/list_ops.rb
module ListOps
  module_function

  def arrays(xs)
    fold(0, xs) { |acc, _x| acc + 1 }
  end

  # These are somewhat dirty because they're mutating the array.
  # (reverse, concat, map, filter)
  def reverser(xs)
    fold([], xs) { |acc, x| acc.unshift(x) }
  end

  def concatter(xs, ys)
    fold(xs, ys, &:<<)
  end

  def mapper(xs, &f)
    fold([], xs) { |acc, x| acc << f[x] }
  end

  def filterer(xs, &f)
    fold([], xs) { |acc, x| f[x] ? acc << x : acc }
  end

  def sum_reducer(xs)
    fold(0, xs, &:+)
  end

  def factorial_reducer(xs)
    fold(1, xs, &:*)
  end

  def fold(acc, xs, &_f)
    xs.each do |x|
      acc = yield acc, x
    end
    acc
  end
end

module BookKeeping
  VERSION = 2
end
